### PR TITLE
Pull out input data sets into their own thing to centralize data validation

### PR DESCRIPTION
**Pull request by collijk**
So I spent a bunch of time saturday digging into this library since it's starting to show up in our models with more frequency.  also, cause I'm curious about bayesian analysis and haven't done much of it myself.  

Random thoughts so far:
- The complexity of the model is bigger than what pymc built it's apis for.  I traced through everything that gets put into the ModelVars class for a model for a `consistent` model and there's just a ton of stuff there.
- What get's put in the model is more complex than it needs to be.  Without a ton of work (which I did!) it's really hard to see what the structure of the model is or what's important to it. It's also clear, that the model is hanging on to a bunch of intermediate values that don't actually matter for the final model. 
- I started  from Yongquan's model for LTBI to understand what's going on.  I cut my applied math teeth on solving systems of odes and optimization problems so I thought I had a pretty good idea of how dismod worked.  Was a little right but somewhat wrong.  There's a language of fixed and random effects in the context of a location hierarchy that still is a little hard for me to parse.  
- Further from there, I'd like to understand use cases a bit better.  This clearly is not sitting inside the GBD machinery anymore (and it'd be useful to know why not).  But you have some external users.  And we have some internal use cases.  Can we strip away the hierarchy stuff?  What's the core functionality?  How do we make it as independent as possible from GBD infrastructure?
- I wanted to start out by doing just a straight port to pymc3, since depending on old libraries is dicey and pymc3 has lines into theano for super fast computation.  The sort of moderate understanding I have of what's going on makes it pretty clear that this model is ripe for optimization.  However, pymc3 abandoned it's directed graph approach to model construction in order to get that interface into theano, and rearchitecting the project around that is too hard until I understand what's going on better.

Lot's of things there.  Decided to just dig in and see if we can clarify things first by cleaning up interfaces and building some intermediate abstractions.  My goal here (and for future PRs until I lose interest or get busy with other things) is to clean up the dismod internals without mucking up the current interface too much.  So:

TL;DR:

- Here is an abstraction around the dismod input data.  It provides a single place to do input data validation and fail (or warn) early if the data is inappropriate.  Pulled validation code from other places.
- I put it in a module called  "parameter" because I think a good intermediate api to try to get to is one based around bayesian network of parameters.  The input data set should be a part of some parameters we have info about.  For now it gets pulled directly up into the model data class.
- Still need to rebuild notebooks.  These are great integration tests we should figure out how to automate.

### CODE REVIEW

**Comment by collijk** [comment]
Q: Do you think it's worth sitting down with the mathematical sciences team (or whoever is doing dismod AT) and coordinating?  I'd like to make software like dismod and vivarium a product of the institute in the same way our data/papers are.  We don't want to "pollute the namespace" with our external collaborators when talking about dismod if possible.
